### PR TITLE
Smarter pusher ramp up during front transtitions for standard VTOLS

### PR DESCRIPTION
**Pull request by GaspardBesacier**
## Describe problem solved by this pull request
Before, for standard VTOLs, no matter what pusher throttle was used before, it was reset to zero at the beginning of a front transition and then linearly increased to reach the desired value for fixed wing mode: it was useless, slower and resulted in a waste of energy. 

## Describe your solution
The solution that we implemented is just to make the ramp up start at the previous value assigned to the pusher instead of resetting it to zero, while keeping the same rate of increase: that way, the drone can just accelerate to get to fixed wing mode without having to slow down before.

## Describe possible alternatives
What we implemented was an easy and simple way to improve the behavior of the drone, so we didn't really consider any other solution. There could be smarter ways to do it, maybe using a nonlinear ramp up but we didn't take the time to focus about it.

## Test data / coverage
It was tested on simulation multiple times and worked fine, and since it is just the way that the pusher throttle is computed during front transitions there aren't any corner cases that could cause problems. 

## Additional context
Before modification:
image
After modification (the front transition is at the end of the plot, from around 43s to 50s) :
Capture d’écran du 2022-10-06 16-44-06

### CODE REVIEW

File: src/modules/vtol_att_control/standard.cpp [hunk 1]
```diff
@@ -237,7 +240,9 @@ void Standard::update_transition_state()
 
 		} else if (_pusher_throttle <= _param_vt_f_trans_thr.get()) {
 			// ramp up throttle to the target throttle value
-			_pusher_throttle = _param_vt_f_trans_thr.get() * time_since_trans_start / _param_vt_psher_rmp_dt.get();
+			_pusher_throttle = _pusher_start_trans + _param_vt_f_trans_thr.get()
```
Comment (sfuhrer) [inline src/modules/vtol_att_control/standard.cpp:243]: Can't it just be a max(min(_pusher_throttle, _param_vt_f_trans_thr), ramped_up_pusher_throttle)?

Like I did for tiltrotor: https://github.com/PX4/PX4-Autopilot/blob/7e49147bcf1473311339a45a92b5026e9640c846/src/modules/vtol_att_control/tiltrotor.cpp#L319 (The min() I didn't add there, would in theory though also make sense.

**Review by sfuhrer — COMMENTED** [review]
Thanks for bringing this up! I have a proposal to achieve the same thing a bit simpler, please have a look if it makes sense to you. 
Overall we should try to keep the amount of states as low as possible, so ideally we get around introducing yet another one with _pusher_start_trans .

**Comment by ThomasRigi** [comment]
@sfuhrer If I understand your proposition correctly, it's not exactly the same as what @GaspardBesacier proposes. The ramp up would start later, meaning that the transition would not be quite as fast finished as with @GaspardBesacier's proposal.
DSC_0106

The idea is very similar, and you indeed fix the resetting of the thrust to zero without introducing a new state. And perhaps you even fulfill the literal definition of VT_PSHER_RMP_DT better in the sense that it's always going to be after VT_PSHER_RMP_DT seconds that thrust reaches its max. But I don't think that this literal definition is really in the spirit of the parameter. Rather, I've always interpreted it as a way of defining the **slope** of the curve rather than the time at which you reach the maximum (which were basically the same thing so far) in order to limit the jerk that the mechanical parts experience.

Arguably, with @GaspardBesacier's approach it would make more sense to replace VT_PSHER_RMP_DT with a new parameter VT_PSHER_SLEW = VT_F_TRANS_THR/VT_PSHER_RMP_DT. I.e. specify the slope of the ramp up directly rather than inferring it from the two existing parameters. A bonus of this new parameter would be that you could also easily limit the rate at which the pusher throttle is changed during MC flight, as there are some rather big jumps sometimes with current logic which can't be good for the longevity of the motor.

If we introduce this new parameter, we'd have a simple way of imposing the ramp I think :

float ramped_pusher_throttle = _pusher_throttle + VT_PSHER_SLEW * dt; // careful with units
_pusher_throttle = min(ramped_pusher_throttle, _param_vt_f_trans_thr.get());

(just an outline, probably needs some additional guards e.g. for 0-values of VT_PSHER_SLEW)

**Comment by sfuhrer** [comment]
@ThomasRigi you're absolutely right, with what I propose the ramp always starts at 0 and not the current value. But it already fixes most of the current issue by not completely switching off the pusher for a while, with a simpler and thus less error-prone logic in place than proposed here initially. 

Your proposal with the slew instead of the ramp time looks indeed cleaner to me, I would support that idea.

File: src/modules/vtol_att_control/standard.cpp (hunk 1 above)
Comment (GaspardBesacier) [reply src/modules/vtol_att_control/standard.cpp:245]: It's true that is it quite a shame not to start ramping up from the beginning of the transition and wait idle for a while as you explained @ThomasRigi but on the other hand I don't see how your idea can be implemented without adding a state, which isn't ideal. Indeed, there is no dt already declared that would simplify the computation of the new throttle, and to define one we would need to keep in memory the last timestamp and thus create a global state.

About your proposal to change the parameter into a slope one, it is looks like a good idea indeed ! I just have one concern: the time for rap up parameter is also used in another part of the code, during the back transition, to calculate the throttle for active braking and I don't really see a way to adapt the code to the new VT_PSHER_SLEW.

Comment (ThomasRigi) [reply src/modules/vtol_att_control/standard.cpp:245]: There already is a _dt in https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/vtol_att_control/vtol_type.h#L222

As for the occurrence of the ramp up time for the back transition reverse thrust, I'm sure we can find an equivalent solution using the new VT_PSHER_SLEW parameter.

**Comment by ThomasRigi** [comment]
Looks good to me. If @sfuhrer agrees as well then we just have to tackle the parameter migration (https://github.com/PX4/PX4-Autopilot/blob/main/src/lib/parameters/param_translation.cpp) so that the change 
VT_PSHER_SLEW = VT_F_TRANS_THR/VT_PSHER_RMP_DT
is done automatically upon firmware upgrade.

File: src/modules/vtol_att_control/standard.cpp [hunk 2]
```diff
@@ -231,13 +231,15 @@ void Standard::update_transition_state()
 	}
 
 	if (_vtol_schedule.flight_mode == vtol_mode::TRANSITION_TO_FW) {
-		if (_param_vt_psher_rmp_dt.get() <= 0.0f) {
+		if (_param_vt_psher_slew.get() <= 0.0f) {
```
Comment (sfuhrer) [inline src/modules/vtol_att_control/standard.cpp:234]:
```suggestion
		if (_param_vt_psher_slew.get() <= FLT_EPSILON) {
```

File: src/modules/vtol_att_control/standard.cpp [hunk 3]
```diff
@@ -291,7 +293,8 @@ void Standard::update_transition_state()
 
 		if (time_since_trans_start >= _param_vt_b_rev_del.get()) {
 			// Handle throttle reversal for active breaking
-			float thrscale = (time_since_trans_start - _param_vt_b_rev_del.get()) / (_param_vt_psher_rmp_dt.get());
+			float thrscale = (time_since_trans_start - _param_vt_b_rev_del.get()) * _param_vt_psher_slew.get() /
```
Comment (sfuhrer) [inline src/modules/vtol_att_control/standard.cpp:296]: Isn't it _param_vt_f_trans_thr.get()/ _param_vt_psher_slew.get() ?

File: src/modules/vtol_att_control/standard_params.c [hunk 4]
```diff
@@ -118,14 +118,14 @@ PARAM_DEFINE_FLOAT(VT_B_REV_OUT, 0.0f);
...
 
 /**
- * Pusher throttle ramp up window
+ * Pusher throttle ramp up slew rate
  *
- * Defines the time window during which the pusher throttle will be ramped up linearly to VT_F_TRANS_THR during a transition
- * to fixed wing mode. Zero or negative values will produce an instant throttle rise to VT_F_TRANS_THR.
+ * Defines the slew rate of the throttle during transitions.
+ * Zero or negative values will produce an instant throttle rise to VT_F_TRANS_THR.
  *
```
Comment (sfuhrer) [inline src/modules/vtol_att_control/standard_params.c:125]: add @unit 1/s

File: src/modules/vtol_att_control/standard_params.c [hunk 5]
```diff
@@ -118,14 +118,14 @@ PARAM_DEFINE_FLOAT(VT_B_REV_OUT, 0.0f);
 PARAM_DEFINE_FLOAT(VT_B_REV_DEL, 0.0f);
 
 /**
- * Pusher throttle ramp up window
+ * Pusher throttle ramp up slew rate
  *
- * Defines the time window during which the pusher throttle will be ramped up linearly to VT_F_TRANS_THR during a transition
- * to fixed wing mode. Zero or negative values will produce an instant throttle rise to VT_F_TRANS_THR.
+ * Defines the slew rate of the throttle during transitions.
```
Comment (sfuhrer) [inline src/modules/vtol_att_control/standard_params.c:123]:
```suggestion
 * Defines the slew rate of the puller/pusher throttle during transitions.
```

File: src/modules/vtol_att_control/standard_params.c [hunk 6]
```diff
@@ -118,14 +118,14 @@ PARAM_DEFINE_FLOAT(VT_B_REV_OUT, 0.0f);
...
+ * Zero or negative values will produce an instant throttle rise to VT_F_TRANS_THR.
  *
  * @max 20
```
Comment (sfuhrer) [inline src/modules/vtol_att_control/standard_params.c:126]: I guess @min should be at 0, @max can be left open.

And add a comment that the a value of 0 means that the slew rate limiting is disabled, meaning it will instantly go the the transition throttle.

**Review by sfuhrer — COMMENTED** [review]
Yes the new parameter needs to be adapted in the configs and a param translation implemented  (see https://github.com/PX4/PX4-Autopilot/blob/main/src/lib/parameters/param_translation.cpp).

File: src/modules/vtol_att_control/standard.cpp [hunk 7]
```diff
@@ -231,13 +231,15 @@ void Standard::update_transition_state()
...
-		if (_param_vt_psher_rmp_dt.get() <= 0.0f) {
+		if (_param_vt_psher_slew.get() <= 0.0f) {
 			// just set the final target throttle value
 			_pusher_throttle = _param_vt_f_trans_thr.get();
 
-		} else if (_pusher_throttle <= _param_vt_f_trans_thr.get()) {
+		} else if (_pusher_throttle < _param_vt_f_trans_thr.get()) {
 			// ramp up throttle to the target throttle value
-			_pusher_throttle = _param_vt_f_trans_thr.get() * time_since_trans_start / _param_vt_psher_rmp_dt.get();
+			_pusher_throttle += _param_vt_psher_slew.get() * _dt;
```
Comment (sfuhrer) [inline src/modules/vtol_att_control/standard.cpp:240]:
```suggestion
			_pusher_throttle = math::min(_pusher_throttle + _param_vt_psher_slew.get() * _dt, _param_vt_f_trans_thr.get());
		
```
Otherwise it can grow above _param_vt_f_trans_thr.get() (depending on the slew rate setting and _dt).

File: src/modules/vtol_att_control/standard.cpp (hunk 3 above)
Comment (sfuhrer) [reply src/modules/vtol_att_control/standard.cpp:297]:
```suggestion
			float thrscale = (time_since_trans_start - _param_vt_b_rev_del.get()) * _param_vt_psher_slew.get() *
```
I think.

[the remark this answers is not in the document]
Comment (ThomasRigi) [reply src/modules/vtol_att_control/standard.cpp:297]: It feels counter-intuitive but I think Gaspard's version is correct:

thrscale should be unitless, right ? 

Gaspard's expression: unitless = [time]*[throttle/time] / [throttle]
Your expression unitless =/= [time]*[throttle/time] * [throttle] = [throttle²]

It's kind of twisting the mind as [throttle] is sort of unitless in the first place...

File: src/lib/parameters/param_translation.cpp [hunk 8]
```diff
@@ -250,5 +250,16 @@ bool param_modify_on_import(bson_node_t node)
 		}
 	}
 
+	// 2022-11-08 (stable release v1.13.1): translate VT_F_TRANS_THR/VT_PSHER_RMP_DT -> VT_PSHER_SLEW)
```
Comment (ThomasRigi) [inline src/lib/parameters/param_translation.cpp:253]: This PR is not part of v1.13.1.

File: src/lib/parameters/param_translation.cpp [hunk 9]
```diff
@@ -250,5 +250,16 @@ bool param_modify_on_import(bson_node_t node)
 		}
 	}
 
+	// 2022-11-09: translate VT_F_TRANS_THR/VT_PSHER_RMP_DT -> VT_PSHER_SLEW)
+	{
+		if (strcmp("VT_PSHER_RMP_DT", node->name) == 0) {
+			strcpy(node->name, "VT_PSHER_SLEW");
+			float _param_vt_f_trans_thr = param_find("VT_F_TRANS_THR");
+			node->d = _param_vt_f_trans_thr / node->d;
```
Comment (ThomasRigi) [inline src/lib/parameters/param_translation.cpp:258]: This fails to build because of implicit double-promotion. I think if you replace `float` with `double` it should work.

File: src/modules/vtol_att_control/standard.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ThomasRigi) [reply src/modules/vtol_att_control/standard.cpp:297]: another way to see that @GaspardBesacier's expression keeps current behaviour is to just use the identity _param_vt_psher_slew = _param_vt_f_trans_thr / _param_vt_psher_rmp_dt.

However, it does feel weird to have _param_vt_f_trans_thr show up in code that is responsible for the back transition. So I would propose to ditch the intermediary `thrscale` and calculate `_pusher_throttle` directly as follows:
`_pusher_throttle = math::constrain((time_since_trans_start - _param_vt_b_rev_del.get()) * _param_vt_psher_slew.get(), 0.0f, _param_vt_b_trans_thr.get())`

**Comment by ThomasRigi** [comment]
@sfuhrer @GaspardBesacier can you take a look at my last proposition in https://github.com/PX4/PX4-Autopilot/pull/20394#discussion_r1020061302 ? If we all agree on this I think we can move on and get this PR merged.

@GaspardBesacier I think you have to rebase on main to solve the conflict in param_translation.cpp. Another parameter was added to the file in the meantime.

@sfuhrer anything to do about the failing jenkins check ? It's also failing on main.

File: src/modules/vtol_att_control/standard.cpp (hunk 7 above)
Comment (sfuhrer) [reply src/modules/vtol_att_control/standard.cpp]: Sorry I wrote it wrong, it should be 
_pusher_throttle = math::min(**_param_vt_f_trans_thr.get()** + _param_vt_psher_slew.get() * _dt, _param_vt_f_trans_thr.get());
, no? Otherwise it will not increase throttle at all during the front transition...

File: src/modules/vtol_att_control/standard.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/vtol_att_control/standard.cpp]: Looks correct now, let's forget about the intermediate step that caused all the confusion :upside_down_face:

**Review by sfuhrer — APPROVED** [review]
Looks good to me.

**Comment by sfuhrer** [comment]
>anything to do about the failing jenkins check ? It's also failing on main.

Let's not worry about that one now, it's broken since a long while.
